### PR TITLE
Allow custom id property for Source + id-related fixes

### PR DESCRIPTION
**Pull request by VictorVelarde**
Fixes
- Fix interactivity styles on GeoJSON, where no cartodb_id is available

Features (internal, not currently exposed)
- Allow specifying a custom id property for a MVTLayer (in viewportFeaturesGenerator), being that a number but also now a string. That's not used right now for SQLSource, as there it is a hardcoded 'cartodb_id', but it is required if we consider another arbitrary tile sources (eg BQ), as it is the key to gather different pieces in tiles with good viewport calculations

Refactor
- I did also some small refactors, doc headers and debug examples renames

### CODE REVIEW

File: examples/_debug/source/geojson/noCartoDbId.html [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        // style
+        const style = carto.viz.style.colorBins('value', {
+          size: 30,
+          palette: 'ORYEL',
+          viewport: true
+        });
+
+        // geojson source (external, to allow custom options)
+        const geojsonSource = new carto.viz.source.GeoJSON(data);
+        // const geojsonSource = new carto.viz.source.GeoJSON(data, { uniqueIdProperty: 'name' });
```
Comment (padawannn) [inline examples/_debug/source/geojson/noCartoDbId.html:66]: For this example, should you uncomment line 66 and comment line 65?

File: src/viz/interactivity/viewport-features/ViewportFeaturesGenerator.ts [hunk 2]
```diff
@@ -58,43 +152,18 @@ export class ViewportFeaturesGenerator {
...
-        return selectPropertiesFrom(feature.properties as Record<string, unknown>, properties);
-      })
-      .flat();
+    return this._deckLayer.state.tileset.selectedTiles;
   }
 
   private getMVTViewportFilteredFeatures(selectedTiles: ViewportTile[], properties: string[]) {
     const currentFrustumPlanes = this.getViewport().getFrustumPlanes();
-    const featureCache = new Set<number>();
+    const featureCache = new Set<string>(); // don't assume just number (the most common use case, eg cartodb_id)
```
Comment (padawannn) [inline src/viz/interactivity/viewport-features/ViewportFeaturesGenerator.ts:166]: Could be better new Set<string|number>() ? what do you think?

File: src/viz/interactivity/viewport-features/ViewportFeaturesGenerator.ts [hunk 3]
```diff
@@ -146,51 +215,88 @@ export class ViewportFeaturesGenerator {
     return isInside;
   }
 
-  private getSelectedTiles(): ViewportTile[] {
-    if (!this.deckLayer || !this.deckLayer.state) {
-      return [];
+  private getFeatureId(feature: GeoJSON.Feature): string {
+    if (!feature.properties) {
```
Comment (padawannn) [inline src/viz/interactivity/viewport-features/ViewportFeaturesGenerator.ts:219]: We cannot get a feature without properties but with feature id?

File: src/viz/interactivity/viewport-features/ViewportFeaturesGenerator.ts [hunk 4]
```diff
@@ -146,51 +215,88 @@ export class ViewportFeaturesGenerator {
...
 }
 
+// #region Private complements
 interface ViewportFeaturesGeneratorOptions {
   uniqueIdProperty?: string;
 }
 
 interface InsideViewportCheckOptions {
-  featureCache: Set<number>;
+  featureCache: Set<string>;
```
Comment (padawannn) [inline src/viz/interactivity/viewport-features/ViewportFeaturesGenerator.ts:297]: Could be better Set<string|number>?

**Review by padawannn — CHANGES_REQUESTED** [review]

File: examples/_debug/source/geojson/noCartoDbId.html (hunk 1 above)
Comment (VictorVelarde) [reply examples/_debug/source/geojson/noCartoDbId.html:66]: Well, it's a different use case... The main fix is the first one, using a geojson file without 'cartodb_id' (that was working wrong on interactivity). Adding a uniqueProperty to GeoJSON is indeed an option I'm now considering if it makes sense to expose or not.

File: src/viz/interactivity/viewport-features/ViewportFeaturesGenerator.ts (hunk 3 above)
Comment (VictorVelarde) [reply src/viz/interactivity/viewport-features/ViewportFeaturesGenerator.ts:219]: That's a good point. I had assumed it "contained" in the properties, but I guess we should also consider the feature.id option

File: src/viz/interactivity/viewport-features/ViewportFeaturesGenerator.ts (hunk 2 above)
Comment (VictorVelarde) [reply src/viz/interactivity/viewport-features/ViewportFeaturesGenerator.ts:166]: That's the idea, but with an explicit conversion from number (integer expected) always to string. Let me have a look at it, to see if I can refactor it

**Comment by VictorVelarde** [comment]
Closing in favour off https://github.com/CartoDB/web-sdk/pull/128 (with updated lock file)

**Pull request closed without merging**